model/role.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"io"
+	"regexp"
+)
+
+const (
+	SYSTEM_USER_ROLE_ID              = "system_user"
+	SYSTEM_ADMIN_ROLE_ID             = "system_admin"
+	SYSTEM_POST_ALL_ROLE_ID          = "system_post_all"
+	SYSTEM_POST_ALL_PUBLIC_ROLE_ID   = "system_post_all_public"
+	SYSTEM_USER_ACCESS_TOKEN_ROLE_ID = "system_user_access_token"
+
+	TEAM_USER_ROLE_ID            = "team_user"
+	TEAM_ADMIN_ROLE_ID           = "team_admin"
+	TEAM_POST_ALL_ROLE_ID        = "team_post_all"
+	TEAM_POST_ALL_PUBLIC_ROLE_ID = "team_post_all_public"
+
+	CHANNEL_USER_ROLE_ID  = "channel_user"
+	CHANNEL_ADMIN_ROLE_ID = "channel_admin"
+
+	ROLE_NAME_MAX_LENGTH = 64
+)
+
+type Role struct {
+	Id            string   `json:"id"`
+	Name          string   `json:"name"`
+	DisplayName   string   `json:"display_name"`
+	Description   string   `json:"description"`
+	Permissions   []string `json:"permissions"`
+	SchemeManaged bool     `json:"scheme_managed"`
+}
+
+type RolePatch struct {
+	Permissions *[]string `json:"permissions"`
+}
+
+func (role *Role) ToJson() string {
+	if b, err := json.Marshal(role); err != nil {
+		return ""
+	} else {
+		return string(b)
+	}
+}
+
+func RoleFromJson(data io.Reader) *Role {
+	decoder := json.NewDecoder(data)
+	var role Role
+	err := decoder.Decode(&role)
+	if err == nil {
+		return &role
+	} else {
+		return nil
+	}
+}
+
+func RoleListToJson(r []*Role) string {
+	b, err := json.Marshal(r)
+	if err != nil {
+		return ""
+	} else {
+		return string(b)
+	}
+}
+
+func RoleListFromJson(data io.Reader) []*Role {
+	decoder := json.NewDecoder(data)
+	var roles []*Role
+	err := decoder.Decode(&roles)
+	if err == nil {
+		return roles
+	} else {
+		return nil
+	}
+}
+
+func (r *RolePatch) ToJson() string {
+	b, err := json.Marshal(r)
+	if err != nil {
+		return ""
+	} else {
+		return string(b)
+	}
+}
+
+func RolePatchFromJson(data io.Reader) *RolePatch {
+	decoder := json.NewDecoder(data)
+	var o RolePatch
+	err := decoder.Decode(&o)
+	if err == nil {
+		return &o
+	} else {
+		return nil
+	}
+}
+
+func (o *Role) Patch(patch *RolePatch) {
+	if patch.Permissions != nil {
+		o.Permissions = *patch.Permissions
+	}
+}
+
+var validRoleNameCharacters = regexp.MustCompile(`^[a-z0-9_]+$`)
+
+func IsValidRoleName(roleName string) bool {
+	if len(roleName) <= 0 || len(roleName) > ROLE_NAME_MAX_LENGTH {
+		return false
+	}
+
+	if !validRoleNameCharacters.MatchString(roleName) {
+		return false
+	}
```

Comment:
A non-regex way to do the same thing:

```go
if strings.TrimLeft(roleName, "abcdefghijklmnopqrstuvwxyz0123456789_") != "" {
```

It doesn't really matter here, but I was curious how the performance would compare:

```
BenchmarkRegex-8   	 2000000	      1000 ns/op
BenchmarkTrim-8    	10000000	       210 ns/op
```